### DISCUSSION TITLE
Eigenvalue analysis of waveguide in 1D periodic structures

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#24436] [Q&A General] [answered]
Hi,

Is it possible to do output eigenvalues of a waveguide in a periodic domain? (1D domain with periodic boundary conditions)

I have attached a description of the problem below. E and rho are the material properties, k and omega the wavenumber and frequency. I would like to output the wavenumbers for a frequency sweep and create a dispersion. 

[Reference .pdf](https://github.com/idaholab/moose/files/11523078/Reference.pdf)

Thanks!

**Comment by lindsayad** [comment]
@cticenhour any thoughts on this?

**Comment by cticenhour** [comment, marked as answer]
Hi @obelix333 - there is an example I made for the calculation of a fundamental eigenvalue for 2D waveguide profiles (circular, coaxial, rectangular, etc.) in the electromagnetics module that you potentially could adapt to your scenario. Here's the link to a discussion of the benchmark, which compares MOOSE EM module results to analytical solutions and an older FEM code: https://mooseframework.inl.gov/benchmarks/WaveguideEigenvalue.html

The input files associated with this are located in the moose repository at: `moose/modules/electromagnetics/test/tests/benchmarks/eigenvalue_problems`

Take a look through this and the input files, and feel free to ask me more questions about it.

**Reply by ghost** [reply]
Thanks @cticenhour.
I will go through this example. Is it possible to output the matrices used to solve the eigen value problem in moose?

**Reply by GiudGiud** [reply]
Does `-ksp_view_mat` do what you want?
There's also a petsc option to dump it to csv. 
You ll want to look at the petsc manual for more on these

**Reply by ghost** [reply]
I aim to use the coefficients matrices to solve a second order eigenvalue problem in MATLAB. 
I tried using `-ksp_view_mat`, but it outputs a bunch of lines in console. Is it possible to export it to csv?

**Reply by GiudGiud** [reply]
yeah there's an option to dump it the matlab format actually.
I dont recall, please search the PETSC manual. Or maybe @oanaoana knows

**Reply by ghost** [reply]
I will look in the PETSC manual. Thanks @GiudGiud

**Reply by ghost** [reply]
Hi,

After going through the PETSc manaul, I found out the options for saving matrices and vectors to a .m file (to be read in MATLAB). However, is it possible to identify which kernel the matrices belong to?

**Reply by GiudGiud** [reply]
All the contributions from all the kernels are agglomerated.

If you want to look at just one contribution, you ll have to use the vector_tag and matrix_tag parameters. Though it will only get you the residual and/or the diagonal of the Jacobian. You can look at it as a field then using the TagVector/MatrixAux to fill an auxvariable

**Reply by ghost** [reply]
Thanks! I don't think that would help me solve for eigenvalues, or would it? I might be missing something, being new to the MOOSE framework.

**Reply by GiudGiud** [reply]
Oh no that won’t be enough.
If you develop, you can use a matrix tag then output the matrix tag to output the matrix to csv

**Comment by ghost** [comment]
Thank you so much @GiudGiud.
